File: src/main/java/it/frob/sleighidea/SleighBraceMatcher.java
// SPDX-License-Identifier: Apache-2.0

package it.frob.sleighidea;

import com.intellij.lang.BracePair;
import com.intellij.lang.PairedBraceMatcher;
import com.intellij.psi.PsiFile;
import com.intellij.psi.tree.IElementType;
import it.frob.sleighidea.psi.SleighTypes;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class SleighBraceMatcher implements PairedBraceMatcher {
    @Override
    public BracePair @NotNull [] getPairs() {
        return new BracePair[]{
                new BracePair(SleighTypes.LBRACE, SleighTypes.RBRACE, true),
                new BracePair(SleighTypes.LPAREN, SleighTypes.RPAREN, true),
                new BracePair(SleighTypes.LBRACKET, SleighTypes.RBRACKET, true)
        };
    }

    @Override
    public boolean isPairedBracesAllowedBeforeType(@NotNull IElementType lbraceType,
                                                   @Nullable IElementType contextType) {
        return false;
    }

    @Override
    public int getCodeConstructStart(PsiFile file, int openingBraceOffset) {
        return openingBraceOffset;
    }
}

File: src/main/kotlin/it/frob/sleighidea/SleighBraceMatcher.kt
// SPDX-License-Identifier: Apache-2.0

package it.frob.sleighidea

import com.intellij.lang.PairedBraceMatcher
import com.intellij.lang.BracePair
import it.frob.sleighidea.psi.SleighTypes
import com.intellij.psi.PsiFile
import com.intellij.psi.tree.IElementType

class SleighBraceMatcher : PairedBraceMatcher {

    override fun getPairs(): Array<BracePair> = arrayOf(
        BracePair(SleighTypes.LBRACE, SleighTypes.RBRACE, true),
        BracePair(SleighTypes.LPAREN, SleighTypes.RPAREN, true),
        BracePair(SleighTypes.LBRACKET, SleighTypes.RBRACKET, true)
    )

    override fun isPairedBracesAllowedBeforeType(lbraceType: IElementType, contextType: IElementType?): Boolean = false

    override fun getCodeConstructStart(file: PsiFile, openingBraceOffset: Int): Int = openingBraceOffset
}